libraries/AP_Generator/AP_Generator_IE_650_800.cpp

```diff
@@ -0,0 +1,163 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_Generator_IE_650_800.h"
+
+#if GENERATOR_ENABLED
+
+#include <stdio.h>
+
+// Update fuelcell, expected to be called at 20hz
+void AP_Generator_IE_650_800::update()
+{
+    if (_uart == nullptr) {
+        return;
+    }
+
+    const uint32_t now = AP_HAL::millis();
+
+   // Read any available data
+    uint32_t nbytes = MIN(_uart->available(),30u);
+    while (nbytes-- > 0) {
+        int16_t c = _uart->read();
+        if (c < 0) {
+            // Nothing to decode
+            continue;
+        }
+
+        if (!decode(c)) {
+            // Sentence not yet valid, don't assign state and output
+            continue;
+        }
+
+        // Successfully decoded a new valid sentence
+        // Update internal Fuel cell state
+        _state.state = (enum FUEL_CELL_STATE)_parsed.state;
+        _state.err_code = _parsed.err_code;
+        set_failsafe_mask(_parsed.err_code);
+
+        // Update variables to be returned to front end
+        _output.fuel_remain_pct = _parsed.tank_pct * 0.01;
+
+        // Invert bat remaining percent to match AP_BattMonitor convention
+        _output.consumed_mah = 100.0f - _parsed.battery_pct;
+
+        // This unit does not report voltage, always report 1 volt
+        _output.voltage = 1;
+
+        _state.last_time_ms = now;
+
+        // Check if we should notify gcs off any change of fuel cell state
+        check_status();
+
+        update_frontend();
+
+        _state.healthy = (now - _state.last_time_ms) < HEALTHY_TIMEOUT_MS;
+    }
+}
+
+// Process characters recieved and build sentence for IE 650/800 W
+void AP_Generator_IE_650_800::decode_latest_term()
+{
+    // Null terminate and move onto the next term
+    _term[_term_offset] = 0;
+    _term_offset = 0;
```

Comment:
Doesn't need changing, but I would have used zero-indexing on _term_number and moved this incrementing and resetting of _term_offset to the bottom of decode_latest_term.